smile/socket_interface.py

```diff
@@ -0,0 +1,160 @@
+#emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
+#ex: set sts=4 ts=4 sw=4 et:
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See the COPYING file distributed along with the smile package for the
+#   copyright and license terms.
+#
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+
+from state import CallbackState
+from clock import clock
+
+import sys
+import socket
+
+_sockets = {}
+
+
+def init_socket_outlet(uniq_ID, server, port):
+    """Sets up a socket that allows for TCP messaging.
+
+    A *SocketPush* state will use a pre-initialized Socket to send out a
+    marker to the specific server and port. Once created, the socket will be
+    added to a dictionary of sockets (_sockets) where the key is a string joined
+    by an *_*. EX **markersocket_127.0.0.1_1234**.
+
+    Parameters
+    ----------
+    unique_id : string
+        Unique identifier for this socket. Used only to differenciate this
+        socket from other sockets initialized with this function.
+    server : string
+        Host name. For a local TCP process, use the string "localhost"
+    port : int
+        A port number for the server.
+
+    Returns a Socket that is to be used in conjunction with the *SocketPush*
+    state. Will return None if the connection could not be established.
+    """
+
+
+    global _sockets
+
+    unique_identifier = "_".join([uniq_ID, server, str(port)])
+
+    if unique_identifier in _sockets.keys():
+        return _sockets[unique_identifier]
+
+    else:
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+
+        try:
+            rtn = sock.connect((server, port))
+            _sockets[unique_identifier] = sock
+            return _sockets[unique_identifier]
+
+        except:
+            sys.stdout.write("[ERROR  ] [SOCKET      ] Unable to establish a " +
+                                      "connection for the socket at server %s and the port %i\n" % (server, port))
```

Comment:
This got tabbed over farther than I intended. We may also want to move the part after the % to the next line. 
```suggestion
                            "connection for the socket at server %s and the port %i\n" % (server, port))
```